tsdb/shard.go

```diff
@@ -1786,136 +1821,193 @@ func (fs *MeasurementFieldSet) IsEmpty() bool {
-		b, err := fs.marshalMeasurementFieldSet()
+		b, err := fs.marshalMeasurementFieldSetNoLock()
 		if err != nil {
-			return true, err
+			return true, fmt.Errorf("failed marshaling fields for %s: %w", fs.path, err)
 		}
 		if b == nil {
 			// No fields, file removed, all done
 			return true, nil
 		}
 		if _, err := fd.Write(b); err != nil {
-			return true, err
+			return true, fmt.Errorf("failed saving fields to %s: %w", path, err)
 		}
 		return false, fd.Sync()
 	}()
-	if err != nil || isEmpty {
-		return
+	if err != nil {
+		return err
+	} else if isEmpty {
+		// remove empty file
+		return os.RemoveAll(fs.path)
 	}
-	err = fs.renameFile(path)
+	return fs.renameFile(path)
 }
 
-// marshalMeasurementFieldSet: remove the fields.idx file if no fields
-// otherwise, copy the in-memory version into a protobuf to write to
-// disk
-func (fs *MeasurementFieldSet) marshalMeasurementFieldSet() ([]byte, error) {
-	fs.mu.Lock()
-	defer fs.mu.Unlock()
-	if len(fs.fields) == 0 {
-		// If no fields left, remove the fields index file
-		if err := os.RemoveAll(fs.path); err != nil {
-			return nil, err
-		} else {
-			return nil, nil
+// writeChangesToFile: Write a change file for fields.idx
+func (fscm *measurementFieldSetChangeMgr) writeChangesToFile(first writeRequest) {
+	var err error = nil
+	// Put the errorChannel on which we blocked into a slice to allow more invocations
+	// to share the return code from the file write
+	errorChannels := []errorChannel{first.errorReturn}
+	changes := []FieldChanges{first.changes}
+	// On return, send the error to every go proc that send changes
+	defer func() {
+		for _, c := range errorChannels {
+			c <- err
+			close(c)
+		}
+	}()
+	// Do some blocking IO operations before marshalling the changes,
+	// to allow other changes to be queued up and be captured in one
+	// write operation, in case we are under heavy field creation load
+	fscm.mu.Lock()
+	defer fscm.mu.Unlock()
+	fd, err := os.OpenFile(fscm.Path, os.O_CREATE|os.O_APPEND|os.O_SYNC|os.O_WRONLY, 0666)
+	if err != nil {
+		err = fmt.Errorf("failed opening %s: %w", fscm.Path, err)
+		return
+	}
+
+	// ensure file closed
+	defer errors2.Capture(&err, fd.Close)()
+
+	// Read all the pending field and measurement write or delete
+	// requests
+	for {
+		select {
+		case wr := <-fscm.writeRequests:
+			changes = append(changes, wr.changes)
+			errorChannels = append(errorChannels, wr.errorReturn)
+			continue
+		default:
 		}
+		break
+	}
+	// marshal the alice of slices of field changes
+	var b []byte
+	b, err = marshalFieldChanges(changes...)
+	if err != nil {
+		err = fmt.Errorf("failed marshaling changes to %s: %w", fd.Name(), err)
+		return
+	}
+	if b == nil {
+		// No fields, nothing to do but close the file.
+		return
+	}
+	if err = writeSizePlusBuffer(fd, b); err != nil {
+		err = fmt.Errorf("failed writing to %s: %w", fd.Name(), err)
+		return
+	}
+	if err = fd.Sync(); err != nil {
+		err = fmt.Errorf("failed syncing %s: %w", fd.Name(), err)
 	}
```

Comment:
The `fd.Sync()` call is redundant. The `O_SYNC` flag in the `OpenFile` call implicitly syncs to the block device after every write. I don't think it hurts to have it, but it isn't necessary.

Reply:
One less error to check without it.....